stlab/future.hpp

```diff
@@ -1042,24 +1047,33 @@ namespace detail
             , _holds(_remaining)
         {}
 
-        std::atomic_size_t                    _remaining{0};
+        std::atomic_size_t                    _remaining;
         std::vector<future<void>>             _holds;
         std::mutex                            _mutex;
         boost::optional<std::exception_ptr>   _error;
-        size_t                                _index{0};
+        size_t                                _index{std::numeric_limits<size_t>::max()};
         packaged_task<>                       _f;
 
         void failure(std::exception_ptr error) {
-            // we already have a result
-            if (_remaining == 0) {
-                return;
-            }
+            size_t current_remaining = _remaining.load();
+            
+            bool run = false;
+            
             // only the last error is of any interest
-            if (_remaining == 1) {
-                _error = std::move(error);
-                _f();
+            while (current_remaining != 0) { // we already have a result
+                size_t new_remaining = current_remaining - 1;
+                if (new_remaining == 0) {    // this was the last chance
+                    _error = std::move(error); 
+                    run = true;
+                    break;
+                }
+
+                if (_remaining.compare_exchange_strong(current_remaining, new_remaining)) {
```

Comment:
You are right. One line was missing where I set current_ramaining in the loop after the cmpxchg failed to make it correct. But a simpler solutions is possible which is part of my next push.